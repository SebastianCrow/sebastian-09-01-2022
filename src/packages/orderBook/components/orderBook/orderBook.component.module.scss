@import "../../../../styles/variables";
@import "../../../../styles/layout";

.container {
  display: flex;
  flex-direction: column;
  height: 100%;

  .header {
    display: flex;
    padding: $spacing (2 * $spacing);

    .title {
      flex-shrink: 0;
    }
  }

  .spreadContainer {
    display: flex;
    justify-content: center;
    width: 100%;

    @include mobile {
      padding: $spacing (2 * $spacing);
    }
  }

  .tablesContainer {
    display: flex;
    height: 100%;
    overflow: hidden;

    @include mobile {
      flex-direction: column-reverse;
      justify-content: center;
    }

    @include desktop {
      .tableColumn {
        flex: 1;
      }
    }
  }

  .toggleFeedContainer {
    display: flex;
    justify-content: center;
    padding: 2 * $spacing;
  }
}
