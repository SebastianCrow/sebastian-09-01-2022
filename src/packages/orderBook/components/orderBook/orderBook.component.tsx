import React, { FunctionComponent, useCallback } from 'react';
import cns from 'classnames';
import { SpreadHeader } from '../spreadHeader/spreadHeader.component';
import styles from './orderBook.component.module.scss';
import { Text } from '../../../ui/text/text.component';
import { OrderBookTable } from '../orderBookTable/orderBookTable.component';
import {
  useSelectOrderBookAsks,
  useSelectOrderBookBids,
  useSelectOrderBookState,
} from '../../hooks/useSelectOrderBookState.hook';
import { Button } from '../../../ui/button/button.component';
import { useStopObservingProductOnDocumentHidden } from '../../hooks/useStopObservingProductOnDocumentHidden.hook';
import { useStartObservingProduct } from '../../hooks/useStartObservingProduct.hook';
import { useLostConnectionNotification } from '../../hooks/useLostConnectionNotification.hook';

// TODO: Replace all the inline styles with Sass

export const OrderBook: FunctionComponent = () => {
  const { observeProduct } = useStartObservingProduct();
  useStopObservingProductOnDocumentHidden();

  useLostConnectionNotification();

  const { product, connectionStatus } = useSelectOrderBookState();
  console.info('connectionStatus', connectionStatus);
  // TODO: Move inside OrderBookTable?
  const bids = useSelectOrderBookBids();
  const asks = useSelectOrderBookAsks();

  const toggleFeed = useCallback(() => {
    observeProduct(product === 'Bitcoin' ? 'Ethereum' : 'Bitcoin');
  }, [observeProduct, product]);

  return (
    <div className={styles.container}>
      <div className={styles.header}>
        <div className={styles.title}>
          <Text>Order Book</Text>
        </div>
        <div className={cns(styles.spreadContainer, 'desktop')}>
          <SpreadHeader />
        </div>
      </div>
      <div className={styles.tableContainer}>
        <div className={styles.tableColumn}>
          <OrderBookTable priceInfoList={bids} priceDataType="bids" />
        </div>
        <div className={cns(styles.spreadContainer, 'mobile')}>
          <SpreadHeader />
        </div>
        <div className={styles.tableColumn}>
          <OrderBookTable priceInfoList={asks} priceDataType="asks" />
        </div>
      </div>
      <div className={styles.toggleFeedContainer}>
        <Button onClick={toggleFeed}>Toggle Feed</Button>
      </div>
    </div>
  );
};
