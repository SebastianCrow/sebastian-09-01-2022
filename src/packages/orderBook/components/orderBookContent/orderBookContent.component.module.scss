@import '../../../../styles/layout';

.tablesContainer {
  display: flex;
  height: 100%;
  overflow-y: auto;

  .loadingOverlay {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
  }

  .tablesContent {
    display: flex;
    width: 100%;
    // `align-items: center` is problematic when the content overflows the flex parent (cannot scroll to part of the content).
    // Internal wrapper with `margin: auto` centers the content and makes it always accessible.
    margin: auto;

    @include mobile {
      flex-direction: column-reverse;
    }

    .spreadContainer {
      display: flex;
      align-items: center;
      justify-content: center;
      padding: $spacing;
    }
  }

  @include desktop {
    align-items: center;

    .tableColumn {
      flex: 1;
    }
  }
}
