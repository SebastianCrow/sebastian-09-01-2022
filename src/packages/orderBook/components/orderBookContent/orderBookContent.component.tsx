import React, { FunctionComponent } from 'react';
import cns from 'classnames';
import { SpreadHeader } from '../spreadHeader/spreadHeader.component';
import styles from './orderBookContent.component.module.scss';
import { Loader } from '../../../ui/loader/loader.component';
import { OrderBookTable } from '../orderBookTable/orderBookTable.component';
import { useSelectOrderBookPrices } from '../../hooks/useSelectOrderBookState.hook';

export const OrderBookContent: FunctionComponent = () => {
  const { bids, asks } = useSelectOrderBookPrices();

  const hasData = Boolean(bids && asks);

  return (
    <div className={styles.tablesContainer}>
      {!hasData && (
        <div className={styles.loadingOverlay}>
          <Loader />
        </div>
      )}
      {hasData && bids && asks && (
        <div className={styles.tablesContent}>
          <div className={styles.tableColumn}>
            <OrderBookTable priceInfoList={bids} priceDataType="bids" />
          </div>
          <div className={cns(styles.spreadContainer, 'mobile')}>
            <SpreadHeader />
          </div>
          <div className={styles.tableColumn}>
            <OrderBookTable priceInfoList={asks} priceDataType="asks" />
          </div>
        </div>
      )}
    </div>
  );
};
