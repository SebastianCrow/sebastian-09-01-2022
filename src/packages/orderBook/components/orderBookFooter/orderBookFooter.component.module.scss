@import '../../../../styles/variables';

.toggleFeedContainer {
  display: flex;
  justify-content: center;
  padding: 2 * $spacing;
}
