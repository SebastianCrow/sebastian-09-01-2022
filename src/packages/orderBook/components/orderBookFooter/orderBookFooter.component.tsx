import React, { FunctionComponent, useCallback } from 'react';
import styles from './orderBookFooter.component.module.scss';
import { Button } from '../../../ui/button/button.component';
import { useSelectOrderBookState } from '../../hooks/useSelectOrderBookState.hook';
import { Product } from '../../services/network/orderBookNetwork.types';
import { FormattedMessage } from 'react-intl';

interface OrderBookFooterProps {
  observeProduct: (product: Product) => void;
}

export const OrderBookFooter: FunctionComponent<OrderBookFooterProps> = ({
  observeProduct,
}) => {
  const { product, connectionStatus } = useSelectOrderBookState();

  const toggleFeed = useCallback(() => {
    observeProduct(product === 'Bitcoin' ? 'Ethereum' : 'Bitcoin');
  }, [observeProduct, product]);

  return (
    <div className={styles.toggleFeedContainer}>
      <Button onClick={toggleFeed} disabled={connectionStatus === 'error'}>
        <FormattedMessage id="Toggle Feed" />
      </Button>
    </div>
  );
};
