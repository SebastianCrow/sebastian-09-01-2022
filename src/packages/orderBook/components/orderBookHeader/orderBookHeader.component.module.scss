@import '../../../../styles/variables';
@import '../../../../styles/layout';

.header {
  display: flex;
  align-items: center;
  min-height: 21px;
  padding: $spacing (2 * $spacing);

  .title {
    position: absolute;
  }

  .spreadContainer {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
  }
}
