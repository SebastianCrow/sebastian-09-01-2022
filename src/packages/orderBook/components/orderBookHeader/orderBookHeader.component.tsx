import React, { FunctionComponent } from 'react';
import { Text } from '../../../ui/text/text.component';
import cns from 'classnames';
import { SpreadHeader } from '../spreadHeader/spreadHeader.component';
import styles from './orderBookHeader.component.module.scss';

export const OrderBookHeader: FunctionComponent = () => {
  return (
    <div className={styles.header}>
      <div className={styles.title}>
        <Text variant="title">Order Book</Text>
      </div>
      <div className={cns(styles.spreadContainer, 'desktop')}>
        <SpreadHeader />
      </div>
    </div>
  );
};
