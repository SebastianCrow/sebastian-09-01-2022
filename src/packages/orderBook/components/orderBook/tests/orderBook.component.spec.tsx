import React from 'react';
import { getOrderBookMockStore } from '../../../../../tests/mocks/getOrderBookMockStore';
import { OrderBookContent } from '../../orderBookContent/orderBookContent.component';
import { OrderBookHeader } from '../../orderBookHeader/orderBookHeader.component';
import { OrderBookFooter } from '../../orderBookFooter/orderBookFooter.component';
import { OrderBook } from '../orderBook.component';
import { createRendererWithOrderBookStore } from '../../../tests/utils/createRendererWithOrderBookStore.util';

describe('orderBook.component', () => {
  test('renders', () => {
    const renderer = createRendererWithOrderBookStore(<OrderBook />);

    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(OrderBookHeader)).toHaveLength(1);
    expect(renderer.root.findAllByType(OrderBookContent)).toHaveLength(1);
    expect(renderer.root.findAllByType(OrderBookFooter)).toHaveLength(1);
  });

  test('renders with `error` connection status', () => {
    const renderer = createRendererWithOrderBookStore(
      <OrderBook />,
      getOrderBookMockStore({
        connectionStatus: 'error',
      })
    );

    expect(renderer.toJSON()).toMatchSnapshot();
    expect(renderer.root.findAllByType(OrderBookHeader)).toHaveLength(0);
    expect(renderer.root.findAllByType(OrderBookContent)).toHaveLength(0);
    expect(renderer.root.findAllByType(OrderBookFooter)).toHaveLength(0);
  });
});
