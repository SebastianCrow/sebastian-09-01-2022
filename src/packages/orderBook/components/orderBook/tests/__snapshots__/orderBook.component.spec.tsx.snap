// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderBook.component renders 1`] = `
<div
  className="container"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 title primary css-ahj2mt-MuiTypography-root"
      >
        Order Book
      </p>
    </div>
    <div
      className="spreadContainer desktop"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 secondary css-ahj2mt-MuiTypography-root"
      >
        Spread: 
        17.00
         (
        0.05
        %)
      </p>
    </div>
  </div>
  <div
    className="tablesContainer"
  >
    <div
      className="tablesContent"
    >
      <div
        className="tableColumn"
      >
        <div
          className="loadingOverlay"
        >
          <span
            className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary loader css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          >
            <svg
              className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx={44}
                cy={44}
                fill="none"
                r={20.2}
                strokeWidth={3.6}
                style={Object {}}
              />
            </svg>
          </span>
        </div>
      </div>
      <div
        className="spreadContainer mobile"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 secondary css-ahj2mt-MuiTypography-root"
        >
          Spread: 
          17.00
           (
          0.05
          %)
        </p>
      </div>
      <div
        className="tableColumn"
      >
        <div
          className="loadingOverlay"
        >
          <span
            className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary loader css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style={
              Object {
                "height": 40,
                "width": 40,
              }
            }
          >
            <svg
              className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx={44}
                cy={44}
                fill="none"
                r={20.2}
                strokeWidth={3.6}
                style={Object {}}
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="toggleFeedContainer"
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root primary css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="toggle-feed-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Toggle Feed
    </button>
  </div>
</div>
`;

exports[`orderBook.component renders with \`error\` connection status 1`] = `
<div
  className="container"
>
  <div
    className="header"
  >
    <div
      className="title"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 title primary css-ahj2mt-MuiTypography-root"
      >
        Order Book
      </p>
    </div>
    <div
      className="spreadContainer desktop"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 secondary css-ahj2mt-MuiTypography-root"
      >
        Spread: 
        17.00
         (
        0.05
        %)
      </p>
    </div>
  </div>
  <div
    className="toggleFeedContainer"
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled primary css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="toggle-feed-button"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      Toggle Feed
    </button>
  </div>
</div>
`;
